src/stirling/source_connectors/perf_profiler/notes.md

```diff
@@ -0,0 +1,209 @@
+```
+stack_trace.push_back(rip);
+```
+3. The function return address is always the last thing pushed to the stack before calling a new function. Thus, the function return address is at memory location `rbp + 8` [2]. Read the function return address and add it to the stack trace.
+```
+fn_return_address = *(fp + 8);
+stack_trace.push_back(fn_return_address);
+```
+4. The saved base pointers each, in turn, point to their caller's stack. Find the next caller's base pointer:
+5. If `saved_rbp` is not a valid stack address, exit, else goto step (3).
+
+In pseudo code:
+```
+fp = *rbp;
+stack_trace.push_back(rip);
+while(true) {
+  fn_return_address = *(saved_rbp + 8);
+  stack_trace.push_back(fn_return_address);
+  fp = *fp;
+  if(!valid_stack_address(fp)) {
+    break;
+  }
+}
+```
+
+[1] For this method to work, the program must be compiled using the "frame pointer" convention, i.e. the compiler optimization `-fomit-frame-pointer` (or equivalent) must not be in use.
+
+[2] x86 stacks grow "down" so the bottom of the stack is the highest memory address in the stack.
+
+###### A normal stack
+```
+-------------------- +---------------+
+Caller_0        +--> |  ............ |         # Saved bp_1 points to here.
+                |    +---------------+
+                |    |  ............ |
+                |    +---------------+
+                |    |  return addr. |
+--------------  |    +---------------+
+Caller_1        +--- |  saved bp_1   | <--+    # Saved bp_2 points to here, the stack frame of the caller of the leaf fn.
+                     +---------------+    |
+                     |  ............ |    |
+                     +---------------+    |
+                     |  return addr. |    |
+-------------------- +---------------+    |
+Leaf fn.    rbp -->  |  saved bp_2   | ---+
+                     +---------------+
+                     |  ............ |
+                     +---------------+
+            rsp -->  |  ............ |
+                     +---------------+
+
+```
+__Figure 1.__ A normal stack (i.e. during function execution, i.e. not at the moment of function entry or function exit).
+
+##### The stack at the moment of function entry or function exit
+```
+-------------------- +---------------+
+Caller_0             |  saved bp_0   |
+                     +---------------+
+                     |  ............ |
+                     +---------------+
+                     |  return addr. |
+-------------------  +---------------+
+Caller_1    rbp -->  |  saved bp_1   |   # points to saved bp_0
+                     +---------------+
+                     |  ............ |
+                     +---------------+
+                     |  return addr. |   # This return address will not get picked up into the stack trace.
+-------------------- +---------------+
+Leaf fn.    rsp -->  |  saved bp_2   |   # points to saved bp_1
+                     +---------------+
```

Comment:
This picture didn't match my mental model of how I thought function entry would look. In the case that we sample before function entry is complete, shouldn't the `rbp` be on the stack "lower" than the `rsp`? Meaning we executed the `push`, but not the `mov` instruction of function entry?

Reply:
Will try to explain what I have, and see if it matches your understanding (what you wrote does make sense, and where needed, I'll add some detail).

Your description of the situation seems right. `push` was executed, `mov` was not. I think for a `push` instruction, the stack pointer is decremented, but the convention is that the stack pointer points to the "top" of the stack (not one beyond the top).  Although the previous base pointer is indeed at the top of the stack, that doesn't help the stack trace walker because it cannot use the stack pointer to find the base pointer. Rather, it just reads register `rbp` and that is its base pointer. Here, `rbp` is pointing to the caller's stack frame not the current function's stack frame. For that reason, it immediately walks back one and starts the stack trace with `Caller_0` (i.e. using the illustration in this doc).

Reply:
What you explained makes sense, but put another way here's what I'm having trouble reconciling: If the stack and registers exist as mentioned in this diagram, how would we lose a non leaf symbol when stack walking?

The symptom we see is that the stack walker always jumps 1 layer, and with the way the diagram is currently written it makes me believe we would always lose the leaf symbol (not a layer between). Is that explanation sound?

If not, we don't need to block this doc on me fully understanding this. Even though I understand the concept of the issue, the inner details are still a bit fuzzy for me :)

Reply:
The stack trace we want is this:
```
ip -> return addr. 0 -> return addr. 1
```
But because `ip` points to `push` while `bp` points into the caller's stack, we get the following:

1. Symbol is from the leaf function (the push instruction is in the leaf function), 
2. We skip return addr. 0 (because we pickup return addr. 1 using the base pointer pointing into the caller's stack), and hence have a missing stack frame (the stack trace goes from leaf to the caller's caller).

The stack trace we get is this:
```
ip -> return addr. 1
```

Reply:
Thanks for that explanation! That clears up my confusion 👍